### ISSUE TITLE
Image files uploaded through custom block wysiwyg are removed by garbage collection

### ISSUE THREAD

**Issue by Graham-72** [backdrop/backdrop-issues#1373]
I have just been having a problem with a disappearing image file. After this happened for the third time I have now used a fresh copy with a different file name. I also found that a cron run produced a report saying 'Could not delete temporary file "temporary://DRClogo150a.png" during garbage collection' - an image file that is no longer in use on the site.

This caused me to wonder whether there is a problem with garbage collection and I looked at the contents of the data table 'file_managed'. This contained some records of image files that are no longer used on the site and I would have thought should have been removed on a cron run.

Has anyone else had a problem with a wanted image being removed and unwanted images still being listed in file_managed?

===
~~PR from @serundeputy https://github.com/backdrop/backdrop/pull/1281~~
~~PR from @jenlampton https://github.com/backdrop/backdrop/pull/1512~~
PR from @quicksketch backdrop/backdrop#1520

**Comment by klonos** [comment]
I vaguely remember an issue in d.org about this. Could be wrong though.

Did you check this: https://drupal.stackexchange.com/questions/60445/why-is-drupal-deleting-my-images

**Comment by klonos** [comment]
Found the issue: [Make unused file 'cleanup' configurable [#1399846]](https://www.drupal.org/node/1399846)

It seems that a remedy for this has been incorporated in D8: [Temporary file deletion window is now configurable](https://www.drupal.org/node/2239983). It's basically the same issue I linked to above that was [merged in D8 back in April 2014](https://www.drupal.org/node/1399846#comment-8678483). It was then switched to D7 with a "Patch (to be ported)" status.

**Comment by klonos** [comment]
Here's the UI change (under ```admin/config/media/file-system```)...

Backdrop:
![backdrop-file_system_config](https://cloud.githubusercontent.com/assets/2423362/11451289/fbd9d900-9614-11e5-927d-0818facfd6d0.png)

Drupal 8:
![d8-file_system_config](https://cloud.githubusercontent.com/assets/2423362/11451292/191187b6-9615-11e5-9a96-c6647b5a9deb.png)

**Comment by klonos** [comment]
You're welcome @Graham-72. Guess I've been spending way too much time in the d.org queues back in the days.

**Comment by Graham-72** [comment]
@klonos Are there perhaps two separate issues here? One is that garbage is not always being cleared properly, and the other (which I have now clarified) is that an image that is the only content of a custom block created with CKEditor disappears at the next cron run, even though it had a correct entry in the file_managed table.

**Comment by klonos** [comment]
Well, the issue is summarized into two phrases:

> Core's file fields have very hard-coded assumptions that if a file is no longer used in a file reference field, then it should be deleted. ...

and:

> ... We are unable to add usage for media embedded into WYSIWYG for one big reason: filters lack context about the entity that has the text field being changed/viewed.

**Comment by Graham-72** [comment]
I am sorry but I am still not understanding my particular situation.

With one exception the images I have been adding with WYSIWYG (CKEditor) are remaining i.e. not being deleted.

In the one case where I have created a custom block with its only content being an image, the image is being deleted by some garbage collection process. Is this a filters problem? Perhaps I should make sure there is some other content in the block, even if I then use CSS and set display:none for the extra content. Just wondering what is happening and why.

**Comment by mikemccaffrey** [comment]
Wow, this is a huge problem. I created a bunch of blocks to hold text and images, and all of the images I uploaded through the WYSIWYG are deleted whenever cron runs.

Has anyone thought of ways to get around this? Turning off garbage collection does not seem like an appropriate solution. Perhaps we could just exclude the inline-images directory? Or could we create some sort of registry to track where inline images are uploaded?

**Comment by mikemccaffrey** [comment]
Hmm... looking into this, I might have found the potential problem. When a file is uploaded through the WYSIWYG, an entry is created in file_managed, but the status is set to 0. I did a test where I updated the status to 1 in the database, and cron no longer deletes the image. We should figure out why the WYSIWYG upload is not setting the status flag itself.

**Comment by mikemccaffrey** [comment]
> Core's file fields have very hard-coded assumptions that if a file is no longer used in a file reference field, then it should be deleted. ...

Not is not actually the case! When an entity is created or updated, the filter module looks for files that have been referenced in process text  fields, and then sets their status to permanent so they are not removed:

```
/**
 * Implements hook_entity_insert().
 */
function filter_entity_insert(EntityInterface $entity) {
  $referenced_files_by_field = _filter_get_file_ids_by_field($entity);
  foreach ($referenced_files_by_field as $field => $fids) {
    _filter_record_file_usage($fids, $entity);
  }
}

/**
 * Records file usage of files referenced by processed text fields.
 *
 * Every referenced file that does not yet have the FILE_STATUS_PERMANENT state,
 * will be given that state.
 *
 * @param array $fids
 *   An array of file entity IDs.
 * @param EntityInterface $entity
 *   An entity whose fields to inspect for file references.
 */
function _filter_record_file_usage(array $fids, EntityInterface $entity) {
  foreach ($fids as $fid) {
    $file = file_load($fid);
    if ($file->status !== FILE_STATUS_PERMANENT) {
      $file->status = FILE_STATUS_PERMANENT;
      $file->save();
    }
    file_usage_add($file, 'filter', $entity->entityType(), $entity->id());
  }
}

```
The only reason we are seeing problem in this case is the custom blocks are not entities, so the insert and update hooks are not being called. 

It seems like we need to figure out some way of recording file usage when an images is uploaded through a text field that does not belong to an entity. Other than custom blocks, can anyone think of places where the wysiwyg is used in a non-entity-editing context?

**Comment by quicksketch** [comment]
Agreed this is huge problem.

> The only reason we are seeing problem in this case is the custom blocks are not entities, so the insert and update hooks are not being called. 

Yep, bingo.

> It seems like we need to figure out some way of recording file usage when an images is uploaded through a text field that does not belong to an entity.

To solve the immediate problem, I think we should simply disable the uploading of images in situations where the file usage entries aren't created. This would include places like custom blocks and in the Views UI.

> Other than custom blocks, can anyone think of places where the wysiwyg is used in a non-entity-editing context?

Views UI in custom text blocks and User Signatures (which is on an entity, but is not a field). And if we add Layout Custom Text blocks in https://github.com/backdrop/backdrop-issues/issues/1099 they will have the same problem.

Because File IDs, file usages, and the files themselves are stored in not stored in configuration, they're not exportable across sites. Disabling the uploading is the only thing I can think of that would be an appropriate solution. Site builders could still use images with manually specified URLs (such as those in the /files directory or possibly in the site theme).

**Comment by quicksketch** [comment]
I'll take handling this issue if it sounds like this solution of just disabling the upload would work.

**Comment by mikemccaffrey** [comment]
> To solve the immediate problem, I think we should simply disable the uploading of images in situations where the file usage entries aren't created. This would include places like custom blocks and in the Views UI.

I think that would work, but I'm not sure how you would do it. Is there an easy way to tell the wysiwyg not to include the image upload option, but keep the image url option? And do we need some sort of message to the user explaining why it is not available? Maybe we can form alter the upload form if it is being called from a non-field, and replace all the form elements with a descriptive notice.

**Comment by mikemccaffrey** [comment]
Also, if we are disabling these image uploads instead of somehow fixing them, we should provide some alternative ui for uploading files to the site without having to attach them to some content entity, such as the file management interface I mentioned in #1448.

**Comment by serundeputy** [comment]
I cannot replicate this ...  has it been solved.

* visit: `/admin/structure/block`
* click `Add custom block` : `/admin/structure/block/add`
* Fill in the admin label field
* Upload an image in the `Block content` textarea using the ckeditor button
* Check the `file_managed` table in the DB
* there is an entry for the new image wit `status` 0
* Run cron
* The image entry is still in the `file_managed` table 
* The image is still in the block

Are my steps not reflective of the problem or is this solved?

**Comment by gifad** [comment]
> Are my steps not reflective of the problem ?

You missed one step : Wait for 6 hours (the `BACKDROP_MAXIMUM_TEMP_FILE_AGE`) and then run cron…

**Comment by gifad** [comment]
A workaround is to install [IMCE](https://backdropcms.org/project/imce)
Not the best UX for end users, but does the job for site builders…

**Comment by serundeputy** [comment]
thanks @gifad that was the missing piece.

>The only reason we are seeing problem in this case is the custom blocks are not entities, so the insert and update hooks are not being called.

I'll agree that the functions  `filter_entity_insert`, `_filter_record_file_usage ` are not being called since the block is not an entity, but something is putting in the database in `file_managed` in the first place.

Wherever that is can't we just set the status to 1 there?  I tried to find that place, but I was unsuccessful.

I don't like the idea of removing the image buttons from ckeditor in certain situations, like on the block add form.

Seems like a lot of these issues are leading towards blocks should be fieldable entities?

I'll keep hunting for the place that the `file_manged` record is added to the database.

**Comment by serundeputy** [comment]
This `PR`, https://github.com/backdrop/backdrop/pull/1281 sets the `file_managed` status column to 1.

```php
/**
 * Submit handler for filter_format_editor_image_form().
 */
function filter_format_editor_image_form_submit($form, &$form_state) {
  // Image source overrides file uploads, as the values are emptied out when
  // hidden by JavaScript.
  if (!empty($form_state['values']['attributes']['src'])) {
    $form_state['values']['fid'] = NULL;
    $form_state['values']['attributes']['data-file-id'] = NULL;
  }
  // Convert any uploaded files from the FID values to the src attribute.
  if (!empty($form_state['values']['fid'])) {
    $fid = $form_state['values']['fid'];
    $file = file_load($fid);
    $file->status = 1;
    $file->save();
    $form_state['values']['attributes']['src'] = file_create_url($file->uri);
    $form_state['values']['attributes']['data-file-id'] = $fid;
    unset($form_state['values']['fid']);
  }
}
```

I've added two lines:

```php
    $file->status = 1;
    $file->save();
```

It passes tests, and works for me™, but I don't know the consequences for other managed files by putting this status to 1 here.  I don't know why or how someone can upload an image elsewhere and set the status to 0?

So this could be a bad idea. I'm just not sure.

https://github.com/backdrop/backdrop/pull/1281.patch

~Geoff

**Comment by gifad** [comment]
> So this could be a bad idea. I'm just not sure.

Two issues : 
1. If the user removes the image before saving, or just does not save the page, an orphan file will stay permanently in the database.
2. Anyway, you will create a managed file without a file_usage…

@quicksketch : is it right to assume that "files_managed" is a shortcut for files **managed by entities** ? (or are there any other exceptions ?)

**Comment by docwilmot** [comment]
> I don't like the idea of removing the image buttons from ckeditor in certain situations, like on the block add form.

Agreed. We need to really start working on [outstanding CKeditor issues](https://github.com/backdrop/backdrop-issues/issues/1087) as @Graham-72 has been pleading for so long, especially a [file/image browser](https://github.com/backdrop/backdrop-issues/issues/1307). 

Unfortunately, all the field and JQuery stuff is completely foreign to me, and I cant help there. This is @quicksketch territory.

**Comment by Graham-72** [comment]
@docwilmot thanks for the mention of my plea. Like you I feel helpless when faced with the field and jQuery stuff. At present I am getting along well with @gifad 's contributed module 'CKEditor addons' but in the particular case of the disappearing image I have had to resort to using CSS to make it a background image for a custom block.

**Comment by gifad** [comment]
@Graham-72  : talking about CKEditor, be carefull : Since `class Block extends LayoutHandler`, we are in a CTools occupied territory, and original CKEditor plugins do not work here : they cannot acquire focus !

@serundeputy
> Seems like a lot of these issues are leading towards blocks should be fieldable entities?

I'm not a PHP programmer, and my request of `class BlockText extends Block, Entity ` is silently ignored by the interpreter…
Perhaps we need a new custom block type, which would embed a fieldable entity - think of [Nodeblock](url)

**Comment by gifad** [comment]
Please check sandbox project just above.

**Comment by quicksketch** [comment]
> @quicksketch : is it right to assume that "files_managed" is a shortcut for files managed by entities ? (or are there any other exceptions ?)

There are situations in contrib, such as Webform Submissions, where a file may be attached to something that is not necessarily an entity.

@gifad is right in https://github.com/backdrop/backdrop-issues/issues/1373#issuecomment-193248709, there a a few reasons why we can't just mark all files permanent upon upload.

**Comment by serundeputy** [comment]
Implement `filter_block_insert` to mark image as permanent.
`filter_block_update` in `filter` module

thanks @quicksketch

**Comment by quicksketch** [comment]
Hi @serundeputy, looks like I was mistaken. There is no `hook_block_insert/update`. There is a `hook_block_save()`, but it's only called for the originating module, not all modules. I realized we also have to deal with reusable blocks (saved in `block.custom.[name].json` files) and one-off blocks (saved in the the `layout.layout.[name].json` file).

So to accommodate for both of these, the best place is probably in `hook_layout_presave()`. Loop through each of the blocks in `$layout->content`, parse for files, and mark them as permanent.

Right now the function that does parsing is "private" in Filter module: `_filter_parse_file_fids()`, we'll probably want to make that public by removing the leading underscore, as we're going to need it in other places.

**Comment by gifad** [comment]
Just a question : is there something equivalent to `$entity->original` for blocks ?

**Comment by Graham-72** [comment]
As a mere site designer (most of the time) may I ask whether we are close to solving this one? The other day I wanted to add some images as blocks in a side column and of course they soon disappeared. My (I hope temporary) solution contimues to be to use CSS to make them background images of divs and, of course, to set the height and width of the div.

**Comment by gifad** [comment]
> whether we are close to solving this

Issue [#1076: Add ability to add an existing node to a layout as a block](https://github.com/backdrop/backdrop-issues/issues/1076) has already a working PR, and is scheduled for 1.4.0
And it's a better way to handle (I mean "manage") blocks containing **_content_**.
(since blocks are not going to be entities in a visible future)

**Comment by Graham-72** [comment]
Thanks - great news.

**Comment by serundeputy** [comment]
re: https://github.com/backdrop/backdrop-issues/issues/1373#issuecomment-201064024

Here is what I see in `$layout->content`

<img width="817" alt="screen shot 2016-05-28 at 11 54 23 am" src="https://cloud.githubusercontent.com/assets/6763671/15628273/086273ea-24cb-11e6-84ca-888a0bd07197.png">

I don't see any html/xml strings in there that contain `fid`'s which seems to be what `_filter_parse_file_fids` wants:

```php
/**
 * Parse an HTML snippet for any data-file-id attributes.
 *
 * @param string $text
 *   The partial (X)HTML snippet to load. Invalid markup will be corrected on
 *   import.
 *
 * @return array
 *   An array of all found FIDs.
 */
function _filter_parse_file_fids($text) {
  $dom = filter_dom_load($text);
  $xpath = new DOMXPath($dom);
  $fids = array();
  foreach ($xpath->query('//*[@data-file-id]') as $node) {
    $fids[] = $node->getAttribute('data-file-id');
  }
  return $fids;
}
```

any ideas?

**Comment by quicksketch** [comment]
Thanks @serundeputy! I plan to take a look at this today or this weekend if possible.

**Comment by quicksketch** [comment]
I'm also going to start tagging this issue into the next bug fix release each version until it gets in so that I will check on it prior to each release.

**Comment by Graham-72** [comment]
I am so, so pleased to see this being given priority! :+1:

**Comment by mikemccaffrey** [comment]
So what exactly is the solution that we are considering? Is to parse the text field when/before saving the block, identify any image files that are referenced, and set the file_managed status for those to 1?

**Comment by quicksketch** [comment]
Bumping this to the next release, as we have several other important fixes to get out.

> Is to parse the text field when/before saving the block, identify any image files that are referenced, and set the file_managed status for those to 1?

That is what the current PR attempts to do yes. However I think this is only one of two possible approaches (the other being to simply disallow image uploads through the editor). I think we should do both. The current PR helps on blocks but doesn't address Views or any number of other places where an editor might end up being used.

**Comment by jenlampton** [comment]
> I'll agree that the functions filter_entity_insert, _filter_record_file_usage are not being called since the block is not an entity, but something is putting in the database in file_managed in the first place. Wherever that is can't we just set the status to 1 there? 

What if every image uploaded via ckeditor was added to the `file_managed` table  marked as 'permanent'... UNLESS it was attached to an entity. In the later case the image can first set as 'temporary', and then be marked as 'permanent' by the entity hooks we are already using. These are also the places where we often have more complicated workflows, and users are more likely to give up before saving.

> IMCE contrib module solves the issue by assigning the file usage to the File object itself, thus creating a "self managed" file.

If every image uploaded via ckeditor was added to the `file_managed` table **with 'usage' pointing at itself** and marked as 'permanent'... that would also give us a tool to determine which of those files were used on (other) entities and which were not.

**Comment by docwilmot** [comment]
how about:
```
function filter_layout_presave($layout) {
  $fids = array();
  // Parse each BlockText block for file ids
  foreach ($layout->content as $content) {
    if (is_a($content, 'BlockText')) {
      $block_content = $content->settings['content'];
      $block_fids = _filter_parse_file_fids($block_content);
      if (isset($block_fids)) {
        $fids = array_merge($fids, $block_fids);
      }
    }
  }

  // Save each fid usage using 'file' as the type like IMCE does.
  foreach ($fids as $fid) {
    $file = file_load($fid);
    if ($file->status !== FILE_STATUS_PERMANENT) {
      $file->status = FILE_STATUS_PERMANENT;
      $file->save();
    }
    file_usage_add($file, 'filter', 'file', $file->fid());
  }
}
```

**Comment by jenlampton** [comment]
that only solves the layout problem. we also need views, and everywhere else formatted text can happen (both in core and contrib).

**Comment by docwilmot** [comment]
Yes but that does solve the layout problem does it not?

**Comment by jenlampton** [comment]
Well, I don't know, I haven't tested it :) But it would be best to fix the bug in all the places, not just one place. And with code that would not need to be copied/repeated in all those places. Doing something like @serundeputy's approach would fix it in all places.

**Comment by docwilmot** [comment]
See @quicksketch comment in https://github.com/backdrop/backdrop-issues/issues/1373#issuecomment-201064024. @serundeputy's method won't work.

But to make this more generic, we could instead do something along the lines of:
```
function filter_entityless_thing_usage_save($text_from_the_thing) {
  $fids = _filter_parse_file_fids($text_from_the_thing);

  // Save each fid usage using 'file' as the type like IMCE does.
  foreach ($fids as $fid) {
    $file = file_load($fid);
    if ($file->status !== FILE_STATUS_PERMANENT) {
      $file->status = FILE_STATUS_PERMANENT;
      $file->save();
    }
    file_usage_add($file, 'filter', 'file', $file->fid());
  }
}
```

And call that function in a `filter_layout_presave()` and `views_presave()` and `signature_presave()` etc. (I made up those last two of course).

Point is though, this layout issue is most pressing. This approach is likely the most logical way to do this without major restructuring.

**Comment by jenlampton** [comment]
That comment was to the approach you are describing. Changing the permanent status only when blocks/layotus are saved.

My approach in https://github.com/backdrop/backdrop-issues/issues/1373#issuecomment-244232735 would fix all the places we have this issue.

**Comment by jenlampton** [comment]
Here's @serundeputy's version (predates my approach, but is the same thing) https://github.com/backdrop/backdrop-issues/issues/1373#issuecomment-192923946

**Comment by jenlampton** [comment]
Ah @docwilmot I think I found the reason that approach was dismissed.  On the PR nate closed in pointing to [this comment](https://github.com/backdrop/backdrop-issues/issues/1373#issuecomment-193248709) as the reason. 

Let's address those items: 

> 1. If the user removes the image before saving, or just does not save the page, an orphan file will stay permanently in the database.

Is this really a problem? It's certainly a far far smaller problem than the one we have now.

> 2. Anyway, you will create a managed file without a file_usage…

Is this a problem? If so, why?

**Comment by docwilmot** [comment]
Oh great, was typing a whole epistle to explain what you just posted.

But see my second compromise approach above re `filter_entityless_thing_usage_save()`. THis is not in fact *just* for layouts, but allows any text from an editor to save its images to the `file_usage` table as a "self-managed" file. We'd simply have to call that function on the text.

**Comment by jenlampton** [comment]
> allows any text from an editor to save its images to the file_usage table as a "self-managed" file.

YES!!  this was what I mentioned in my post above too!  Are we agree-fighting? :)

**Comment by docwilmot** [comment]
> Are we agree-fighting?

That is now one of my favorite expressions. Going to pick an agree fight with the wife as soon as she gets home. And we sort of are agree-fighting except for the 

>UNLESS it was attached to an entity 

comment (which implies that non-entity files will just be plain marked permanent, which is not best). We still need a new `filter_entityless_thing_usage_save()` and we'd need to call it separately every time a entity-free editor saves something.

**Comment by jenlampton** [comment]
> which implies that non-entity files will just be plain marked permanent, which is not best

Oh, that's not what I meant. For entities images uploaded would be initially marked temporary, and then marked permanent in the _save handlers -- just like they are today. 

The only change for entries is that for more "complex workflows" that setting could be changed initially so that they aren't marked permanent. (Since the maintainer knows enough to handle the images specifically).

**Comment by docwilmot** [comment]
I'm probably a bit slow here, but bear with me. Your suggestion is:
- mark every file uploaded through CKEditor as permanent (I'm assuming in `filter_format_editor_image_form_submit()`) except those which are attached to entities
- then those attached to entities can be marked permanent later

This to me still implies that those files NOT attached to an entity (such as the culprits which started this issue) WOULD be marked as permanent?? And still leaves us with the issues raised in https://github.com/backdrop/backdrop-issues/issues/1373#issuecomment-193248709, no?

**Comment by jenlampton** [comment]
Yes, exactly.  Hence my question: why is that a problem?

**Comment by jenlampton** [comment]
> This to me still implies that those files NOT attached to an entity (such as the culprits which started this issue) WOULD be marked as permanent??

Oh also one more point of clarification - anything that is not an entity still **could** mark things as temporary initially and then mark them as permanent in their _save routines, if they wanted to. It just wouldn't be required as in the previous approaches mentioned above.  (And core would already be handling that for all entities.)

**Comment by docwilmot** [comment]
Just for the record, the method @quicksketch and I suggested would avoid both problems of orphaned files and files without usage. 

> why is that a problem?

Those files would never be deleted by Backdrop though not in use. And having a usage record ensures that files that are in use are not deleted. AFAIK.

**Comment by jenlampton** [comment]
> Just for the record, the method @quicksketch and I suggested would avoid both problems of orphaned files and files without usage.

but it doesn't fix this bug (anywhere other than blocks)

> Those files would never be deleted by Backdrop though not in use

And why is that a problem?

> having a usage record ensures that files that are in use are not deleted. AFAIK.

Ah, not deleted by garbage collection (like this bug) but deleted manually if the file is used in more than one place. Okay, I getcha. But I also agree about having the self-referenced file entry!  So, problem solved? :)

**Comment by docwilmot** [comment]
> but it doesn't fix this bug (anywhere other than for blocks/layouts)

Final try, see suggestion re `filter_entityless_thing_usage_save()`. I'll leave this here for now.

**Comment by jenlampton** [comment]
see https://github.com/backdrop/backdrop-issues/issues/1373#issuecomment-244506924 

If you won't exlpain it, how about a pull request where you add the function `filter_entityless_thing_usage_save` in a way that works for all entityless things in both core and contrib?

**Comment by docwilmot** [comment]
Ahh, sorry, my apologies, I hadnt noticed your edits to that comment. Glad you've found a solution though. Will await the PR.

**Comment by jenlampton** [comment]
the proof will be in the pudding.... if I can figure out the code! :)

**Comment by jenlampton** [comment]
Okay, let's see if this PR makes more sense than me talking around in circles. Here's what's changed:

* Files uploaded via rich-text editor marked as `permanent` (by default).
* For some thingies(tm) (like entities, as demonstrated in this PR) the files can be marked as `temporary` initially, and then marked as permanent when the thingy(tm) is saved.

I might work on one more change for this PR that also has files stored as `temporary` for blocks and then saved as `permanent` by layouts, as recommended by @docwilmot and @quicksketch  above. 

By combining both these approaches, we 1) never delete any image-uploads by garbage collection (win!), and 2) don't generate too many garbage files on disk, as 99% of use-cases for uploading images are covered in core by marking them as temporary first, then permanent at time of save, now with an API contrib can follow.

**Comment by klonos** [comment]
>>Are we agree-fighting?

>That is now one of my favorite expressions.

:) mine too.

**Comment by klonos** [comment]
I don't fully understand what you guys are talking about, but it seems that we are close to sorting this. Good news!

**Comment by quicksketch** [comment]
@serundeputy and I some implementation feedback posted in the PR: https://github.com/backdrop/backdrop/pull/1512

Here's an example of a problem that is created by the "permanent-first" approach:

I create a new custom block (Admin -> Structure -> Custom blocks). I upload an image through CKEditor. Now, I delete the image from the rich text editor body (not through the dialog). From a user-perspective, I would think that I had now "deleted" that file, because it's no longer in the editor. But that is not the case, the file would have been saved to disk and marked as permanent. Whether I save the block or just close the window, I now have permanent cruft on my server. Or worse, I've uploaded something incriminating, personal, or offensive and **I absolutely need to delete that file. But I can't.** (Speaking as a site administrator of course. As a developer or sysadmin I could use SFTP and MySQL to delete it).  There is no tool within Backdrop to remove it, as we don't have a page for Admin -> Content -> Files. 

I think this lack of a file management interface (and thus permanent accumulation of cruft and/or undesirable files) is my biggest problem with the "permanent-first" approach.

WordPress saves every image as it is uploaded as permanent, but it also gives you a tool for managing those images. I don't think it does any attempt at cleaning up files, so if you delete a post it leaves any files that were uploaded within it. In Backdrop, it cleans up files associated with content. We have a fundamentally (at this point) different approach to file management. In WordPress, files are most frequently managed as their own concept, separate from posts. In Backdrop, files are most frequently considered to be a part of something else such as an entity.

So if we were to pivot to files being independently managed instead of generally being part of something else, I would be a lot more comfortable with this approach, but that is a much much larger problem than the bug we're trying to cover here.

Fundamentally, I think the "permanent-first" vs "temporary-first" is our main hang up on this issue. I'll try to break it down into Pros/Cons:


Permanent first:
- Pros:
  - Easy for developers to have file saving "just work" in that they are always permanent and no special saving routine needs to be done.
  - Files can be uploaded by users everywhere where a filtered text element is used.

- Cons:
  - Leaves file cruft on disk when content is unsaved or deleted.
  - Users have no way to delete sensitive or unwanted files.

Temporary first:
- Pros:
  - Files are cleaned up on unsaved or deleted non-entities (no cruft).

- Cons:
  - Requires developers to write code to handle association of files and non-entities.
  - Users are unable to upload files in the editor on unsupported places (though they could still use a hard-coded path).

I can see why "permanent-first" is appealing, it's easier from a DX standpoint and gives users more functionality. However, it also introduces a major bug in that we'd literally have no way for users to delete orphaned files. That could be a problem from a disk-usage standpoint, but I'm much more concerned about the problem of deleting sensitive/unwanted files.

The "temporary-first" approach takes more work from developers and provides generally less functionality (unless developers accommodated for all situations where a filtered text area was used). It's only advantage is that doesn't intentionally introduce a new bug.

So at this point it comes down to how important we think the "cruft bug" is. If it's important enough, we should bite the bullet and do the harder implementation for developers and reduce functionality available to users (until it can be supported later). Or if it's not important, we can plow forward knowing we've made the bug and can resolve it later when we add the ability to manage files (#1448).

**Comment by quicksketch** [comment]
Sorry for the epic above. :stuck_out_tongue: 

I drafted up a PR that allows uploads to be disabled on a per-field basis, and then disabled the uploads on all text areas in Backdrop *except* text.module and block.module, assuming we'll fix block.module in this issue. This PR would be most applicable if we stick with temporary-first files, but it might be useful in permanent-first as well.

This PR doesn't fix or even touch on marking files permanent or not, it just gives the ability to enable/disable uploads on a per-field basis, so it could be disabled in situations where uploading is either difficult to support or not appropriate.

PR: https://github.com/backdrop/backdrop/pull/1520

**Comment by jenlampton** [comment]
Before I get into the details, I wanted to put up a reply specifically addressing Backdrop's philosophy.

Permanent first:
* DX win - easy to form alter things 
* UX win - image uploads work everywhere
* teeny tiny extreme minority issue with files used in uncommon places being undeletable

Temporary first:
* UX nightmare (files deleted by garbage collection), or
* UX confusion (why is there no image upload allowed here?)
* DX nightmare - trying to add inline image support elsewhere is painful

**Comment by jenlampton** [comment]
Now, I understand that we have a problem with deleting images. I think we have this problem anyway, say a block is deleted, we don't delete the files that were referenced from that block. Isn't that a bigger issue than the image someone accidentally almost posted into a views header somewhere?

We need to solve the how-to-delete/manage-files issue regardless of how we handle this particular issue. What if, like @quicksketch mentioned, we committed to putting in an image/media center like WP has. We need to do that anyway! Maybe we should put it on the roadmap for Backdrop 1.6?  

Here's a proposed solution:
* save images as permanent everywhere (except below)
* save images in entities as temporary first
* save images in blocks as temporary first
* put Media on the roadmap for 1.6 https://github.com/backdrop/backdrop-issues/issues/1448

**Comment by docwilmot** [comment]
Third option:
- disable uploading of images on all fields by default (nate's PR)
- enable images just on layout blocks and implement the approach in https://github.com/backdrop/backdrop-issues/issues/1373#issuecomment-244500430 to mark them permanent.
- document why.

Resolves @quicksketch concerns (and mine) re leftover undeletable cruft, and prevents @jenlampton concerns re UX/DX nightmares. Contrib can enable uploading in fieldAPI as in @quicksketch suggestion if they implement a method to mark files permanent.

Keep in mind that this issue has been delayed almost a year; the idea that Media will 'go in' is uncertain.

**Comment by jenlampton** [comment]
> disable uploading of images on all fields by default (nate's PR)

There's no way we're disabling inline images on content. I hope that's not what you meant? Ifnot, then what's the difference between your approach and the second one I mentioned?

**Comment by jenlampton** [comment]
> Keep in mind that this issue has been delayed almost a year

Neither of the solutions I mentioned would be blocked on Media going in (or rather, our version of file entity, which I am calling Media b/c WP did). Both fix this problem in 1.5.  Well, mine fixes it, the other decreases its impact.

**Comment by jenlampton** [comment]
> what's the difference between your approach and the second one I mentioned

Oh I see, nate's PR doesn't actually fix the bug. For some reason I thought it did...  yes, let's add that into option #2!

**Comment by jenlampton** [comment]
> and prevents @jenlampton concerns re UX/DX nightmares. Contrib can enable uploading in fieldAPI as in @quicksketch suggestion if they implement a method to mark files permanent.

Actually, those are my concerns about both UX and DX:
* UX: people wonder why they can't upload images in some places and can in others.
* DX: contrib needs to learn some complicated way to mark the files as permanent in their own save handlers.

**Comment by quicksketch** [comment]
Here's a plan that (might) make everyone happy. Essentially @docwilmot's last suggestion (https://github.com/backdrop/backdrop-issues/issues/1373#issuecomment-244653382) but with the following amendment:

- disable uploading of images on all fields by default (nate's PR)
- enable images just on layout blocks and implement the approach in #1373 (comment) to mark them permanent.
- document why. (and allow for other usages a la carte)
- Amendment: When the full media browser and file entities interface is added, switch to permanent-first and enable uploads everywhere again. Because users will have the ability to manage/delete their files as stand-alone content, instead of being bound to other things like blocks or even entities.

This results in minimal changes in the API, fixes the "99%" use-case, punts on the 1% use-case until we add full image browser and file management.

**Comment by jenlampton** [comment]
Works for me, but I would like to get those UX and DX problems solved asap. Let's leave the file-entity-media-whatever on the map for 1.6 and see how it progresses.

Also, Should we open a follow-up bug report about the cruft files not being deleted from disk when blocks that have had images uploaded are deleted?  I know how you guys feel about cruft! :)

**Comment by quicksketch** [comment]
> Should we open a follow-up bug report about the cruft files not being deleted from disk when blocks that have had images uploaded are deleted?

Yeah... even with this solution we'll *still* have cruft in some situations. We'll fix the situation where "files are uploaded but the block is not saved", and the "files are uploaded but then removed from the block before saving", but we still may not have fixed the "block was deleted that referenced files". It depends on how far we go with the fix. @docwilmot's comment https://github.com/backdrop/backdrop-issues/issues/1373#issuecomment-244500430 fixes addition but doesn't completely solve a block being deleted (either through the UI or through configuration sync). Let's make the follow-up once we see what this looks like.

I'll work on merging @docwilmot's suggested approach into my PR as an additional commit.

**Comment by jenlampton** [comment]
@quicksketch and you'll need to switch the default upload value to `FALSE` instead of `TRUE` or we won't fix the bug. Don't forget that part ;)

**Comment by jenlampton** [comment]
> Or worse, I've uploaded something incriminating, personal, or offensive and **I absolutely need to delete that file. But I can't.** (Speaking as a site administrator of course. As a developer or sysadmin I could use SFTP and MySQL to delete it). There is no tool within Backdrop to remove it, as we don't have a page for Admin -> Content -> Files.

If this is really a concern, then we should open that follow-up bug report about that asap. I don't want that thought to get lost in all our other file-entity-media-browser excitement.

**Comment by quicksketch** [comment]
I adjusted the first commit in my PR to default to FALSE for `#editor_uploads` and specified `TRUE` for custom blocks. We'll need tests for the new functionality of course (recording file usages and permanent state).

Current PR now with two commits: https://github.com/backdrop/backdrop/pull/1520.

**Comment by jenlampton** [comment]
kudos to baby steps :)

**Comment by quicksketch** [comment]
Thanks @docwilmot for the PR review. I'll get another update out today hopefully.

**Comment by quicksketch** [comment]
I made the follow-up for removing file usages at https://github.com/backdrop/backdrop-issues/issues/2137, because I need to reference it in a `@todo` in the PR.

**Comment by quicksketch** [comment]
PR at https://github.com/backdrop/backdrop/pull/1520 now has added tests to confirm the saving of file usage records for both "Custom blocks" (block module) and "Text blocks" (layout module). I ended up not trying to centralize the adding of file usage records because I found that the code was slightly different between Layout module and Block module when saving file usages. Layout module can do all file usages across the entire layout at once, whereas Block module will do them one block at a time.

At first look, I thought that code would "double-fire" for saving file usages, but on further checking Layout module won't affect the blocks owned by Block module, because they are a different class (BlockLegacy) than the one-off text blocks used by Block (BlockText).

I made a few TODO lines in this PR, as we've discussed there are some situations in which cruft can still accumulate. The follow-up at https://github.com/backdrop/backdrop-issues/issues/2137 might be handled directly, but its possible that it will be handled as a side-effect of central file management system in https://github.com/backdrop/backdrop-issues/issues/1448.
